# Concatenating DataFrames

Sometimes, instead of merging DataFrames horizontally (joining columns), we want to stack them vertically (adding rows). This is very common when dealing with time-series data or logs that are split across multiple files, where each file represents a different time period but shares the exact same columns. This is called concatenation.

In [15]:
# Import pandas for data processing
import pandas as pd

### Loading Data

Let's load the College Scorecard data from different years. Notice that since the files are large and have many columns, we pass `low_memory=False` to pandas to ensure it reads the column types correctly without running out of memory while guessing.

In [27]:
# Read College Scorecard data for academic years starting in 2009, 2010, and 2012
# Setting low_memory=False ensures that pandas reads all rows to infer data types correctly
df_2009 = pd.read_csv("datasets/college-scorecard/MERGED2009_10_PP.csv", low_memory=False)
df_2010 = pd.read_csv("datasets/college-scorecard/MERGED2010_11_PP.csv", low_memory=False)
df_2012 = pd.read_csv("datasets/college-scorecard/MERGED2012_13_PP.csv", low_memory=False)


Preview the first 3 rows of the 2010 dataset. Notice the huge number of columns (1977!):

In [28]:
# Display the top 3 rows
df_2010.head(3)

,UNITID,OPEID,OPEID6,INSTNM,CITY,STABBR,ZIP,ACCREDAGENCY,INSTURL,NPCURL,...,OMAWDP8_NOTFIRSTTIME_POOLED_SUPP,OMENRUP_NOTFIRSTTIME_POOLED_SUPP,OMENRYP_FULLTIME_POOLED_SUPP,OMENRAP_FULLTIME_POOLED_SUPP,OMAWDP8_FULLTIME_POOLED_SUPP,OMENRUP_FULLTIME_POOLED_SUPP,OMENRYP_PARTTIME_POOLED_SUPP,OMENRAP_PARTTIME_POOLED_SUPP,OMAWDP8_PARTTIME_POOLED_SUPP,OMENRUP_PARTTIME_POOLED_SUPP
0,100654,00100200,1002,Alabama A & M University,Normal,AL,35762,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,100663,00105200,1052,University of Alabama at Birmingham,Birmingham,AL,35294-0110,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,100690,02503400,25034,Amridge University,Montgomery,AL,36117-3553,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Let's check the number of rows (entries) in each dataset before combining them.

In [29]:
# Check the length (number of records) of each DataFrame
print(len(df_2010))
print(len(df_2009))
print(len(df_2012))

7414
7149
7793


### Basic Concatenation

We pass a list of the DataFrames to `pd.concat()`. This sticks them together row by row. If you sum the numbers from above (7414 + 7149 + 7793), you get 22356, which matches the total row count of the concatenated DataFrame!

In [30]:
# Put the dataframes in a list so we can pass them to the concat function
frames = [df_2009, df_2010, df_2012]

# Concatenate them vertically. Pandas automatically aligns columns with matching names.
pd.concat(frames)

,UNITID,OPEID,OPEID6,INSTNM,CITY,STABBR,ZIP,ACCREDAGENCY,INSTURL,NPCURL,...,OMAWDP8_NOTFIRSTTIME_POOLED_SUPP,OMENRUP_NOTFIRSTTIME_POOLED_SUPP,OMENRYP_FULLTIME_POOLED_SUPP,OMENRAP_FULLTIME_POOLED_SUPP,OMAWDP8_FULLTIME_POOLED_SUPP,OMENRUP_FULLTIME_POOLED_SUPP,OMENRYP_PARTTIME_POOLED_SUPP,OMENRAP_PARTTIME_POOLED_SUPP,OMAWDP8_PARTTIME_POOLED_SUPP,OMENRUP_PARTTIME_POOLED_SUPP
0,100654,00100200,1002,Alabama A & M University,Normal,AL,35762,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,100663,00105200,1052,University of Alabama at Birmingham,Birmingham,AL,35294-0110,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,100690,02503400,25034,Amridge University,Montgomery,AL,36117-3553,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,100706,00105500,1055,University of Alabama in Huntsville,Huntsville,AL,35899,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,100724,00100500,1005,Alabama State University,Montgomery,AL,36104-0271,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7788,47691101,4205801,42058,SAE Institute of Technology San Francisco,Emeryville,CA,94608,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7789,47701101,10145905,1459,Strayer University-Bloomington Campus,Bloomington,MN,554311411,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7790,47702001,10145903,1459,Strayer University-Schaumburg Campus,Schaumburg,IL,601735081,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7791,47702002,10145902,1459,Strayer University-Downers Grove Campus,Downers Grove,IL,605151169,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### Concatenation with Keys

A problem with simple concatenation is that we lose the origin of the data. To fix this, `pd.concat` provides a `keys` parameter which creates a hierarchical index (MultiIndex) that labels the origin of each row.

In [31]:
# Concatenate the DataFrames but pass a 'keys' list to label which year each chunk belongs to.
# Note: the keys below label the chunks as 2011, 2012, and 2013 respectively.
pd.concat(frames, keys=['2011','2012','2013'])

UNITID     OPEID  OPEID6  \
2011 0       100654  00100200    1002   
     1       100663  00105200    1052   
     2       100690  02503400   25034   
     3       100706  00105500    1055   
     4       100724  00100500    1005   
...             ...       ...     ...   
2013 7788  47691101   4205801   42058   
     7789  47701101  10145905    1459   
     7790  47702001  10145903    1459   
     7791  47702002  10145902    1459   
     7792  47702003  10145906    1459   

                                              INSTNM           CITY STABBR  \
2011 0                      Alabama A & M University         Normal     AL   
     1           University of Alabama at Birmingham     Birmingham     AL   
     2                            Amridge University     Montgomery     AL   
     3           University of Alabama in Huntsville     Huntsville     AL   
     4                      Alabama State University     Montgomery     AL   
...                                              ...            ...    ...   
2013 7788  SAE Institute of Technology San Francisco     Emeryville     CA   
     7789      Strayer University-Bloomington Campus    Bloomington     MN   
     7790       Strayer University-Schaumburg Campus     Schaumburg     IL   
     7791    Strayer University-Downers Grove Campus  Downers Grove     IL   
     7792           Strayer University-Aurora Campus         Aurora     IL   

                  ZIP  ACCREDAGENCY  INSTURL  NPCURL  ...  \
2011 0          35762           NaN      NaN     NaN  ...   
     1     35294-0110           NaN      NaN     NaN  ...   
     2     36117-3553           NaN      NaN     NaN  ...   
     3          35899           NaN      NaN     NaN  ...   
     4     36104-0271           NaN      NaN     NaN  ...   
...               ...           ...      ...     ...  ...   
2013 7788       94608           NaN      NaN     NaN  ...   
     7789   554311411           NaN      NaN     NaN  ...   
     7790   601735081           NaN      NaN     NaN  ...   
     7791   605151169           NaN      NaN     NaN  ...   
     7792   605066220           NaN      NaN     NaN  ...   

           OMAWDP8_NOTFIRSTTIME_POOLED_SUPP  OMENRUP_NOTFIRSTTIME_POOLED_SUPP  \
2011 0                                  NaN                               NaN   
     1                                  NaN                               NaN   
     2                                  NaN                               NaN   
     3                                  NaN                               NaN   
     4                                  NaN                               NaN   
...                                     ...                               ...   
2013 7788                               NaN                               NaN   
     7789                               NaN                               NaN   
     7790                               NaN                               NaN   
     7791                               NaN                               NaN   
     7792                               NaN                               NaN   

           OMENRYP_FULLTIME_POOLED_SUPP  OMENRAP_FULLTIME_POOLED_SUPP  \
2011 0                              NaN                           NaN   
     1                              NaN                           NaN   
     2                              NaN                           NaN   
     3                              NaN                           NaN   
     4                              NaN                           NaN   
...                                 ...                           ...   
2013 7788                           NaN                           NaN   
     7789                           NaN                           NaN   
     7790                           NaN                           NaN   
     7791                           NaN                           NaN   
     7792                           NaN                           NaN   

